java/src/main/native/src/ColumnViewJni.cpp

```diff
@@ -257,7 +257,11 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnView_reduce(JNIEnv *env, jclas
     auto col = reinterpret_cast<cudf::column_view *>(j_col_view);
     auto agg = reinterpret_cast<cudf::aggregation *>(j_agg);
     cudf::data_type out_dtype = cudf::jni::make_data_type(j_dtype, scale);
-    return release_as_jlong(cudf::reduce(*col, agg->clone(), out_dtype));
+    return release_as_jlong(
+        cudf::reduce(*col,
+                     std::unique_ptr<cudf::reduce_aggregation>(
+                         dynamic_cast<cudf::reduce_aggregation *>(agg->clone().release())),
```

Comment:
This seems rather unsafe. Can the value `j_agg` be reinterpreted as a `cudf::reduce_aggregation*` earlier and then cloned, rather than reinterpreting as `cudf::aggregation*`, cloning it, and then dynamically casting it to `cudf::reduce_aggregation*`? Still unsafe but might be one cast instead of two. I may be misunderstanding the semantics but I hope something that is either safer or shorter can be done here (and similarly elsewhere).

Reply:
The issue here is that the clone() function (pure virtual) always returns a `std::unique_ptr<aggregation>` so there has to be a cast anyways.  I can make the initial cast to a be a `reduce_aggregation` and then do a second `reinterpret_cast` back to reduce_aggregation after the clone.

Pinging @jrhemstad  for advice.

Reply:
So after further testing here, I think this is the only way to do this.  Any attempt to reinterpret cast directly from the jlong to a `reduce_aggregation` causes failures down the line, though I'm not 100% sure why.

Reply:
So as far as the clone() goes, I believe it's just there to respect the intent of the `unique_ptr`.  We're just being handed a raw pointer from who-knows-where.  

The JNI layer uses the default `aggregation` for creating aggregation (`make_sum_aggregation()`, `make_product_aggregation()`, etc).  I believe that due to the virtual inheritance, you cannot make assumptions about the underlying object structure.  

A quick test here shows, x == y, but x != z.
```
auto x = cudf::make_sum_aggregation().release();
auto y = reinterpret_cast<cudf::reduce_aggregation*>(x);
auto z = dynamic_cast<cudf::reduce_aggregation*>(x);
```

However, in this case, they are all the same.
```
auto x = cudf::make_sum_aggregation<reduce_aggregation>().release();
auto y = reinterpret_cast<cudf::reduce_aggregation*>(x);
auto z = dynamic_cast<cudf::reduce_aggregation*>(x);
```

So the `dynamic_cast` seems to be necessary to handle whatever we might be passed.

Reply:
> but x != z

Is that because the dynamic cast is failing and returning a nullptr? Just trying to make sure I'm interpreting your test case correctly.

Reply:
No, it succeeds but you're getting a different pointer.  I can add the null check.